File: chess.ts/move.ts
import { Field } from './field'
import { Figure } from './figure'

export class RegularMove {
    from: Field;
    to: Field;

    constructor(f: Field, t: Field) {
        this.from = f;
        this.to = t;
    }
}

export class PromotionMove {
    from: Field;
    to: Field;
    figure: Figure;

    constructor(f: Field, t: Field, fi: Figure) {
        this.from = f;
        this.to = t;
        this.figure = fi;
    }
}

export class EnPassantMove {
    from: Field;
    to: Field;
    captured: Field;

    constructor(f: Field, t: Field, c: Field) {
        this.from = f;
        this.to = t;
        this.captured = c;
    }
}

export class CastlingMove {
    from: Field;
    to: Field;
    rookFrom: Field;
    rookTo: Field;

    constructor(f: Field, t: Field, rf: Field, rt: Field) {
        this.from = f;
        this.to = t;
        this.rookFrom = rf;
        this.rookTo = rt;
    }
}

export type Move = RegularMove | PromotionMove | EnPassantMove | CastlingMove

File: chess.py/move.py
from dataclasses import dataclass

from field import Field
from figure import Figure


@dataclass
class Move:
    frm: Field
    to: Field


@dataclass
class RegularMove(Move):
    pass


@dataclass
class PromotionMove(Move):
    figure: Figure


@dataclass
class EnPassantMove(Move):
    captured: Field


@dataclass
class CastlingMove(Move):
    rook_from: Field
    rook_to: Field
